Add SQL election module

evankanderson opened the pull request:
Adds a generic SQL leader election module, as discussed in #3219; it sounds like Sigstore is looking at possibly using this for simplicity.

Fixes #3219

Note that I haven't yet wired this up into Trillian nor added database table declarations -- I'm happy to do that in this PR or a follow-up.  I'm not including this in the changelog until it's wired in, but I wanted to get feedback on how to do that first.

### Checklist

<!---
Go over all the following points, and put an `x` in all the boxes that apply.
Feel free to not tick any boxes that don't apply to this PR (e.g. refactoring may not need a CHANGELOG update).
If you're unsure about any of these, don't hesitate to ask. We're here to help!
-->

- [ ] I have updated the [CHANGELOG](CHANGELOG.md).
- [ ] I have updated [documentation](docs/) accordingly (including the [feature implementation matrix](docs/Feature_Implementation_Matrix.md)).

JAORMX commented on util/election2/sql/election.go:
```diff
@@ -0,0 +1,269 @@
+				e.leaderLock.L.Lock()
+				if leader != e.currentLeader {
+					// Note: this code does not actually care _which_ instance was
+					// elected, it sends notifications on each leadership cahnge.
```
s/cahnge/change/

JAORMX commented on util/election2/sql/election.go:
```diff
@@ -0,0 +1,269 @@
+	}
+	defer tx.Rollback()
+	row := tx.QueryRow(
+		"SELECT leader, last_update FROM leader_election WHERE resource_id = $1",
```
Wouldn't the `$<num>` format be problematic in mysql?

JAORMX commented on util/election2/sql/election.go:
```diff
@@ -0,0 +1,269 @@
+}
+
+func (e *Election) initializeLock(ctx context.Context) error {
+	insert, err := e.db.Prepare("INSERT INTO leader_election (resource_id, leader, last_update) VALUES ($1, $2, $3)")
```
Wouldn't we need to handle the case where more than one replica executes an initialization at the same time? Or is the idea is to crash loop, restart and continue?

roger2hk commented:
/gcbrun

JAORMX commented:
/gcbrun

evankanderson commented:
I fixed the remaining lint issue and added what I think is the right integration hooks.  I'm still working on running the integration tests -- in particular, adding a test that uses the `--election_backend=mysql` option.

evankanderson commented:
/gcbrun

mhutchinson commented on cmd/trillian_log_signer/main.go:
```diff
@@ -47,13 +47,15 @@ import (
 	"github.com/google/trillian/util/election"
 	"github.com/google/trillian/util/election2"
 	etcdelect "github.com/google/trillian/util/election2/etcd"
+	"github.com/google/trillian/util/election2/sql"
```
This comment relates to the package of the new code. I recommend `mysql` over `sql` to be clear at which implementation it is targeting, and for consistency with the flag names used in this file to pick it.

JAORMX commented:
Ah! Linter found something

JAORMX commented:
/gcbrun

phbnf commented on cmd/trillian_log_signer/main.go:
```diff
@@ -143,13 +146,22 @@ func main() {
 	var electionFactory election2.Factory
 	switch {
-	case *forceMaster:
+	case *forceMaster || *electionBackend == "noop":
```
It's not fantastic that we can now control this behavior with two different flags, but I also cannot thing of a better way forward for now. On the long run, we could / should probably deprecate forceMaster. No action required!

phbnf commented on util/election2/mysql/election.go:
```diff
@@ -0,0 +1,280 @@
+}
+
+// Close implements election2.Election
+func (e *Election) Close(ctx context.Context) error {
```
Given this is essentially a wrapper around Resign, this election object could be re-used for a new election, which is slightly at odds with the interface definition. In practice, I don't think anything relies on this behaviour, and there aren't any resources that can be released, so I think it's safe. Does that match your understanding?

evankanderson replied:
Yes.

Do you want me to make the object unusable, to catch errors that might occur by calling other methods after `Close`?

phbnf replied:
If you can think of a way of an elegant way of doing this, sure thing. I don't think it's super super important, it's just some extra care to make sure it's aligned with the interface definition :)

roger2hk commented on util/election2/mysql/election.go:
```diff
@@ -0,0 +1,280 @@
+	return <-done
+}
+
+// WithMastership implements election2.Election
+func (e *Election) WithMastership(ctx context.Context) (context.Context, error) {
```
Having the inclusivity and diversity in mind, may I suggest replacing `master` with another inclusive term?

https://developers.google.com/style/word-list#master

evankanderson replied:
I can rewrite the `election2.Election` interface if you'd like.  😁 

Otherwise, I think those are the only two instances of the word "master" rather than "leader".

evankanderson commented on util/election2/mysql/election.go:
```diff
@@ -0,0 +1,280 @@
+	currentLeader leaderData
+	leaderLock    sync.Cond
+
+	// If a channel is supplied with the cancel, it will be signalled when the election routine has exited.
```
I wrote this before realizing that I could re-use `Resign` for `Close`, and I think one of those did not wait for the error in the first implementation.  I can remove the comment, and change the type from `*chan error` to `chan error` probably...

phbnf replied:
Gotcha - makes sense to me to remove the comment then yes!

vipulagarwal commented:
@evankanderson thanks for this PR as it simplifies Sigstore dependencies. Do we have any plans to merge this PR any soon? Thanks

evankanderson commented:
Sorry, I got distracted by this by life.  If this is still of interest, I can try to revive it after my camping break this week -- maybe by around 14 July or so.
